### PR TITLE
Start exam and submit answer sheet with Timer

### PR DESCRIPTION
**Pull request by yongsokheng**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -36,6 +36,7 @@ gem "devise"
 # gem "capistrano-rails", group: :development
 # Use mysql as the database for Active Record
 gem "bootstrap-sass"
+gem "faker", "1.4.2"
```
Comment (HorNorin) [inline Gemfile]: move to development xD

File: app/controllers/exams_controller.rb [hunk 2]
```diff
@@ -5,15 +5,30 @@ def index
     @exam = Exam.new
     @exams = Exam.all
   end
-  
+
+  def show
+    @exam = Exam.find params[:id]
+    @category= @exam.category
```
Comment (HorNorin) [inline app/controllers/exams_controller.rb]: space

**Comment by HorNorin** [comment]
ok

**Comment by RathanakSreang** [comment]
Ok

File: app/models/exam.rb [hunk 3]
```diff
@@ -3,4 +3,15 @@ class Exam < ActiveRecord::Base
...
   has_many :questions, through: :exams_questions
+
+  accepts_nested_attributes_for :exams_questions
+
+  after_create :create_exams_questions
+
+  private
+  def create_exams_questions
+    question_ids = category.questions.order("RAND()")
+      .limit(category.question_number).map(&:id)
```
Comment (chuanhtuan) [inline app/models/exam.rb]: dùng pluck thay vì map nhé

File: app/models/exam.rb [hunk 4]
```diff
@@ -3,4 +3,15 @@ class Exam < ActiveRecord::Base
   belongs_to :category
   has_many :exams_questions, dependent: :destroy
   has_many :questions, through: :exams_questions
+
+  accepts_nested_attributes_for :exams_questions
+
+  after_create :create_exams_questions
+
+  private
+  def create_exams_questions
```
Comment (chuanhtuan) [inline app/models/exam.rb]: write in a scope

File: app/views/exams/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<% provide :title, @category.name %>
+<div class="row">
+  <div class="col-md-8 col-md-offset-1">
+    <div class="title">
+      <center><b><h2><%= @category.name %></h2></b></center>
+    </div>
+    <%= form_for @exam do |f| %>
+      <%= f.hidden_field :status, value: "Unchecked" %>
```
Comment (chuanhtuan) [inline app/views/exams/show.html.erb]: do not send this attribute from client
plz set in controller

File: app/views/exams/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      <ol>
+        <%= f.fields_for :exams_questions do |ff| %>
+          <li><%= ff.object.question.question %></li>
+          <%= ff.collection_radio_buttons :option_id, ff.object.question.options, 
+                                            :id, :option do |rb| %>
+            <%= rb.label(class: "radio"){rb.radio_button + rb.text} %>
+          <% end %>
+        <% end %>
+      </ol>
+      <% if "Start" == @exam.status %>
```
Comment (chuanhtuan) [inline app/views/exams/show.html.erb]: constants should be define in Setting

File: app/controllers/exams_controller.rb [hunk 7]
```diff
@@ -3,17 +3,43 @@ class ExamsController < ApplicationController
 
   def index
     @exam = Exam.new
-    @exams = Exam.all
+    @exams = Exam.paginate(page: params[:page], per_page: 15)
```
Comment (RathanakSreang) [inline app/controllers/exams_controller.rb]: user can view only his own exam or all?

File: app/controllers/exams_controller.rb [hunk 8]
```diff
@@ -3,17 +3,43 @@ class ExamsController < ApplicationController
...
+    redirect_to exams_url
+  end
+
+  def update
+    @exam = Exam.find params[:id]
+    if @exam.time_out?
+      params[:exam][:status] = Settings.status.unviewed
+    end
+    if @exam.update_attributes exam_params
+      flash[:notice] = t "flashs.success_submit"
```
Comment (RathanakSreang) [inline app/controllers/exams_controller.rb]: if time user can not update exam

File: app/models/exam.rb [hunk 9]
```diff
@@ -3,4 +3,30 @@ class Exam < ActiveRecord::Base
   belongs_to :category
   has_many :exams_questions, dependent: :destroy
   has_many :questions, through: :exams_questions
+
+  accepts_nested_attributes_for :exams_questions
+
+  scope :create_random_question, ->(category){category.questions
+    .limit(category.question_number)
```
Comment (RathanakSreang) [inline app/models/exam.rb]: it seen name of scope  not match  its definition(it not create question)

File: app/controllers/exams_controller.rb (hunk 8 above)
Comment (yongsokheng) [inline app/controllers/exams_controller.rb]: I already check this condition in view exam => show

**Comment by RathanakSreang** [comment]
Ok

**Comment by HorNorin** [comment]
ok

File: app/models/exams_question.rb [hunk 10]
```diff
@@ -2,4 +2,15 @@ class ExamsQuestion < ActiveRecord::Base
   belongs_to :exam
   belongs_to :question
   belongs_to :option
+
+  scope :select_correct_option, ->{joins(:option)
+    .where("correct_option = true")}
+
+  def check_answer
+    if option_id && option.correct_option?
+      "<div class='label label-success'>#{Settings.answer.correct}</div>"
```
Comment (chuanhtuan) [inline app/models/exams_question.rb]: không viết html trong controller và model

File: app/views/exams/_exam.html.erb [hunk 11]
```diff
@@ -1,8 +1,23 @@
...
+  <td><%= exam.created_at.strftime("%d-%m-%Y %I:%M %p") %></td>
   <td><%= exam.category.name %></td>
   <td><%= exam.status %></td>
-  <td><%= exam.test_time %></td>
-  <td><%= exam.result %></td>
-  <td><%= link_to t("buttons.start"), "" %></td>
+  <td><%= exam.category.duration %></td>
+  <td><%= exam.category.question_number %></td>
+  <td>
+    <% if Settings.status.viewed == exam.status %>
```
Comment (chuanhtuan) [inline app/views/exams/_exam.html.erb]: nên viết các hàm exam.viewed? trong model

File: app/models/exams_question.rb [hunk 12]
```diff
@@ -2,4 +2,7 @@ class ExamsQuestion < ActiveRecord::Base
   belongs_to :exam
   belongs_to :question
   belongs_to :option
+
+  scope :select_correct_option, ->{joins(:option)
```
Comment (chuanhtuan) [inline app/models/exams_question.rb]: bad naming
em nên đặt tên là correct_options

File: app/views/exams/_exam.html.erb [hunk 13]
```diff
@@ -1,8 +1,23 @@
...
   <td><%= exam.category.name %></td>
   <td><%= exam.status %></td>
-  <td><%= exam.test_time %></td>
-  <td><%= exam.result %></td>
-  <td><%= link_to t("buttons.start"), "" %></td>
+  <td><%= exam.category.duration %></td>
+  <td><%= exam.category.question_number %></td>
+  <td>
+    <% if exam.viewed? %>
+      <% count_correct_option = exam.exams_questions.select_correct_option.count
```
Comment (chuanhtuan) [inline app/views/exams/_exam.html.erb]: ???

File: app/models/exams_question.rb [hunk 14]
```diff
@@ -2,4 +2,7 @@ class ExamsQuestion < ActiveRecord::Base
   belongs_to :exam
   belongs_to :question
   belongs_to :option
+
+  scope :correct_option, ->{joins(:option)
```
Comment (chuanhtuan) [inline app/models/exams_question.rb]: correct_options

**Pull request merged**